File: src/main/java/j/flyweight/BigCharFactory.java
package j.flyweight;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class BigCharFactory {
  private final Map<String, BigChar> pool = new ConcurrentHashMap<>();
  private static final BigCharFactory singleton = new BigCharFactory();

  private BigCharFactory() {}

  public static BigCharFactory getInstance() {
    return singleton;
  }

  public BigChar getBigChar(char charName) {
    return pool.getOrDefault(String.valueOf(charName), new BigChar(charName));
  }
}

File: src/main/kotlin/k/flyweight/BigCharFactory.kt
/* (C)2022 */
package k.flyweight

import java.util.concurrent.ConcurrentHashMap

object BigCharFactory {
  private val pool: MutableMap<String, BigChar> = ConcurrentHashMap()

  fun getBigChar(charName: Char): BigChar {
    return pool.getOrDefault(charName.toString(), BigChar(charName))
  }
}
